docs/migrations_in_practice.rst

```diff
@@ -0,0 +1,343 @@
+  command; actual code relying on the migration
+
+Once the PRs have both been approved, **merge PR 1**, then set up a private release containing that change. Merging the PR first will prevent migration conflicts with anyone else working in the area, and it's a good idea that anything run on prod is on the master branch. Run your schema migration and management command directly:
+
+    ``cchq <ENV> django-manage --release=<NAME> migrate <APP_NAME>``
+
+Then merge PR 2. The subsequent deploy will run your management command again,
+though it should be very quick this time around, since nearly all data has been
+migrated, and finally the code changes will go live.
+
+The big limitation here is that there’s a gap between the final run of the
+management command and go-live (especially with the variation). Any changes in
+the interim won’t be accounted for. This is sometimes acceptable if you’re
+confident no such changes will have happened (eg, the migration pertains only to
+a custom feature, and we know that project won’t have relevant activity during
+that period).
+
+User Logging Example
+....................
+
+Consider attempting to apply our example user logging migration with a single
+deploy. Make two PRs as described, so they can be merged independently. Then
+while coordinating with the team, merge the first PR, deploy a private release,
+and run the schema migration, then the management command.
+
+The second PR can be merged and go live with the next deploy. This django
+migration will re-run the management command, picking up any new changes since
+it was previously run. In our case, this should be a small enough data set that
+it won’t hinder the deploy. *However*, any changes in the window between that
+run and go-live will not be migrated. To pick up those changes, you can run the
+management command a third time after the deploy, which will ensure all user
+logs have been migrated.
+
+This is still not ideal, since for the period between go-live and this third
+run, there will be missing data in the DB and that data will be in-use in the
+UI. Remember also that third party environments will have the management command
+run only once, on the second deploy (unless we announce this as a required
+maintenance operation), which would mean their data would have a gap in it.
+
+While this single-deploy option is tempting compared to waiting weeks to get out
+a multi-deploy migration, it’s really only suitable for specific situations like
+custom work and unreleased features, where we can be confident the drawbacks are
+insignificant.
+
+
+Best practices for data migrations in Python
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+**Don’t fetch all data at once** - Instead, use an iterator that streams data in
+chunks (note that django queryset’s ``.iter()`` method does not do this). Some
+models have their own performant getters, for others, consider
+``queryset_to_iterator`` for SQL models, ``iter_update`` or ``IterDB`` for couch
```

Comment:
It would be nice to also mention [`DocsIterator`](https://github.com/dimagi/commcare-hq/blob/96506d5a16921469277f7ff02ff7852aade5de00/corehq/util/doc_processor/couch.py#L11-L16) here, which is a fairly new addition. I think it's one of the most robust ways to iterate over all docs of a given doc type.

Caveat: the iteration must be done fully in a single process and cannot be resumed. Use something like `resumable_view_iterator` if you need that power.

Reply:
Oh nice, I wasn't aware of those utils.  Feel free to add those in yourself if you've got the energy, otherwise I'll do so on my next pass when I get to it.